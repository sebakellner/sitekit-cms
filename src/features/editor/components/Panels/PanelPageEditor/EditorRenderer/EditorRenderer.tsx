import type { ComponentPropConfig } from '@features/editor/types/editor.types'
import { editorMap } from '../Editors/editorMap'

type Props = {
  propConfig: ComponentPropConfig
  value: unknown
  onChange: (val: unknown) => void
}

const EditorRenderer = ({ propConfig, value, onChange }: Props) => {
  const title = propConfig.title ?? 'Property'
  const { editor, options } = propConfig

  const EditorComponent = editorMap[editor]

  if (!EditorComponent) return null

  return (
    <EditorComponent
      title={title}
      value={typeof value === 'string' ? value : ''}
      onChange={onChange}
      options={options ?? []}
    />
  )
}

export default EditorRenderer
