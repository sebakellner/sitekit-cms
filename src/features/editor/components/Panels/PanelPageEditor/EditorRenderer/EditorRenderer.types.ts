import type { ComponentPropConfig } from '@features/editor/types/editor.types'

export type EditorRendererProps = {
  propConfig: ComponentPropConfig
  value: unknown
  onChange: (val: unknown) => void
}
