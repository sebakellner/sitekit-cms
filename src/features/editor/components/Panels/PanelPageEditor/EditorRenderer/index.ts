export { default as EditorRenderer } from './EditorRenderer'
export type { EditorRendererProps } from './EditorRenderer.types'
