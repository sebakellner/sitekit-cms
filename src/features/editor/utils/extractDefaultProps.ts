import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
