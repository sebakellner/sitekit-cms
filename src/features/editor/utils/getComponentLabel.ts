import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
