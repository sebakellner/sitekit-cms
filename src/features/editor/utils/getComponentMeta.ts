import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
