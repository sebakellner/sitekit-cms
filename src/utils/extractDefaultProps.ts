import type { ComponentMeta } from '@components/site/types'

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined
): Record<string, unknown> {
  if (!props) return {}
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => [
      key,
      (val as { default: unknown }).default,
    ])
  )
}
