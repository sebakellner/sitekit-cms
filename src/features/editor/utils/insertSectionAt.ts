import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
