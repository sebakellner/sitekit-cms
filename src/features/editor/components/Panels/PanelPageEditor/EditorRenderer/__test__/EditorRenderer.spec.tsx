import { expect, describe, test } from 'vitest'
import { render } from 'test/utils/render'
import EditorRenderer from '../EditorRenderer'
import type { EditorRendererProps } from '../EditorRenderer.types'
import { screen } from '@testing-library/react'
import type { ComponentPropEditorType } from '@features/editor/schemas/componentPropEditor.schema'

const setup = (overrides: Partial<EditorRendererProps> = {}) => {
  const defaultProps: EditorRendererProps = {
    config: {
      title: 'Property 1',
      type: 'string',
      editor: 'text',
      default: 'Default Value',
    },
    value: undefined,
    onChange: () => {},
  }
  const props = { ...defaultProps, ...overrides }
  return <EditorRenderer {...props} />
}

describe('EditorRenderer', () => {
  test('renders TextEditor with default value', () => {
    render(setup())
    const input = screen.getByRole('textbox', { name: 'Property 1' })
    expect(input).toHaveValue('Default Value')
  })

  test('renders TextEditor with provided value', () => {
    render(setup({ value: 'Custom Value' }))
    const input = screen.getByRole('textbox', { name: 'Property 1' })
    expect(input).toHaveValue('Custom Value')
  })

  test('renders SelectEditor with options and default', () => {
    render(
      setup({
        config: {
          title: 'Property 1',
          type: 'enum',
          editor: 'select',
          default: 'opt2',
          options: [
            { label: 'Option 1', value: 'opt1' },
            { label: 'Option 2', value: 'opt2' },
          ],
        },
      })
    )
    const select = screen.getByRole('combobox', { name: 'Property 1' })
    expect(select).toHaveValue('opt2')
    expect(select).toHaveDisplayValue('Option 2')
  })

  test('renders SelectEditor with provided value', () => {
    render(
      setup({
        value: 'opt1',
        config: {
          title: 'Property 1',
          type: 'enum',
          editor: 'select',
          default: 'opt2',
          options: [
            { label: 'Option 1', value: 'opt1' },
            { label: 'Option 2', value: 'opt2' },
          ],
        },
      })
    )
    const select = screen.getByRole('combobox', { name: 'Property 1' })
    expect(select).toHaveValue('opt1')
    expect(select).toHaveDisplayValue('Option 1')
  })

  test('renders ColorPickerEditor with default value', () => {
    render(
      setup({
        config: {
          title: 'Color',
          type: 'string',
          editor: 'colorPicker',
          default: '#fff',
        },
      })
    )
    expect(screen.getByLabelText('Color')).toBeInTheDocument()
  })

  test('renders ListEditor with array value', () => {
    render(
      setup({
        config: {
          title: 'List',
          type: 'enum',
          editor: 'listEditor',
          default: [{ label: 'Item 1' }],
        },
        value: [{ label: 'Item 2' }],
      })
    )
    expect(screen.getByDisplayValue('Item 2')).toBeInTheDocument()
  })

  test('renders nothing for unknown editor', () => {
    render(
      setup({
        config: {
          title: 'Unknown',
          type: 'string',
          editor: 'unknown-editor' as unknown as ComponentPropEditorType,
          default: 'Default',
        },
      })
    )
    expect(screen.queryByText('Unknown')).not.toBeInTheDocument()
  })

  test('renders default title if config.title is not defined', () => {
    render(
      setup({
        config: {
          title: undefined,
          type: 'string',
          editor: 'text',
          default: 'Default Value',
        },
      })
    )
    expect(screen.getByText('Property')).toBeInTheDocument()
  })

  test('should match snapshot for TextEditor', () => {
    const { container } = render(
      setup({
        config: {
          title: 'Text',
          type: 'string',
          editor: 'text',
          default: 'Default',
        },
      })
    )
    expect(container).toMatchSnapshot()
  })

  test('should match snapshot for SelectEditor', () => {
    const { container } = render(
      setup({
        config: {
          title: 'Select',
          type: 'enum',
          editor: 'select',
          default: 'opt1',
          options: [{ label: 'Option 1', value: 'opt1' }],
        },
      })
    )
    expect(container).toMatchSnapshot()
  })

  test('should render the component with the correct props', () => {
    render(
      setup({
        config: {
          title: 'Property 1',
          type: 'string',
          editor: 'text',
          default: 'Default Value',
        },
      })
    )
    expect(screen.getByText('Property 1')).toBeInTheDocument()
  })

  test('should render empty string if value its not a stringu in TextEditor', () => {
    render(
      setup({
        value: 123,
        config: {
          title: 'Property',
          type: 'string',
          editor: 'text',
          default: 'Default value',
        },
      })
    )
    const input = screen.getByRole('textbox', { name: 'Property' })
    expect(input).toHaveValue('')
  })

  test('should render empty array if value is not an array in ListEditor', () => {
    render(
      setup({
        value: 'not-an-array',
        config: {
          title: 'List',
          type: 'enum',
          editor: 'listEditor',
          default: [{ label: 'Item 1' }],
        },
      })
    )
    expect(screen.getByRole('button')).toBeInTheDocument()
  })

  test('should render empty options if options are not provided in SelectEditor', () => {
    render(
      setup({
        config: {
          title: 'Select',
          type: 'enum',
          editor: 'select',
          default: undefined,
          options: undefined,
        },
      })
    )
    const select = screen.getByRole('combobox', { name: 'Select' })
    expect(select).toBeInTheDocument()
    expect(select.querySelectorAll('option').length).toBe(0)
  })

  test('should not render editor if config.editor is not defined', () => {
    render(
      setup({
        config: {
          title: 'Property 1',
          type: 'string',
          editor: undefined as unknown as ComponentPropEditorType,
          default: 'Default Value',
        },
      })
    )
    expect(screen.queryByText('Property 1')).not.toBeInTheDocument()
  })
})
