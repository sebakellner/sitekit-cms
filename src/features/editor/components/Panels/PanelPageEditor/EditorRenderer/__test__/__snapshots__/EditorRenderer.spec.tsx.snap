// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditorPanelRenderer > should render the component with the correct metadata 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-n33ttv"
  >
    <div
      class="chakra-field__root css-1anu6qw"
      data-part="root"
      data-scope="field"
      id="field::«r0»"
      orientation="horizontal"
      role="group"
    >
      <label
        class="chakra-field__label css-84rnco"
        data-part="label"
        data-scope="field"
        for="«r0»"
        id="field::«r0»::label"
      >
        Property 1
      </label>
      <div
        class="chakra-native-select__root css-8sp2e0"
      >
        <select
          class="chakra-native-select__field css-jwpnxr"
          data-part="select"
          data-scope="field"
          id="«r0»"
        >
          <option
            value="option1"
          >
            Option 1
          </option>
          <option
            value="option2"
          >
            Option 2
          </option>
        </select>
        <div
          class="chakra-native-select__indicator css-hydxzs"
        >
          <svg
            class="css-s3mb0o"
            viewBox="0 0 24 24"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
